enhancements/security/runtime-class-scc.md

```diff
@@ -0,0 +1,248 @@
+
+These may be defined in terms of API maturity, or as something else. Initial proposal
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this
+enhancement:
+
+- Maturity levels
+    - [`alpha`, `beta`, `stable` in upstream Kubernetes][maturity-levels]
+    - `Dev Preview`, `Tech Preview`, `GA` in OpenShift
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning),
+or by redefining what graduation means.
+
+In general, we try to use the same stages (alpha, beta, GA), regardless how the functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+#### Examples
+
+These are generalized examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+##### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+
+##### Tech Preview -> GA 
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
+##### Removing a deprecated feature
+
+- Announce deprecation and support policy of the existing feature
+- Deprecate the feature
+
+### Upgrade / Downgrade Strategy
+
+There are two scenerios for both upgrade and downgrade that we have to consider: a user configures a RuntimeClass other than the default, and they do not. In the case of not configuring a runtime class, the user will see no behavior change between upgrades or downgrades.
+
+When a user does specify a runtime class:
+Upgrade: A user should not have a workload that configures a runtime class before this SCC enhancement is accepted and implemented. the MCO will have no additional runtime classes before admins are able to gate runtime classes from users. Thus, before an upgrade, a user's request for a different runtime class would fail. After an upgrade, the request would either succeed or fail, depending if they have permission to use the requested runtime class.
+
+Downgrade: The MCO does not configure any additional RuntimeClasses for versions of Openshift before this SCC enhancement will be complete. Thus, a users workload *may* go from functioning (if they have permission to use the runtime class that is configured), to failing, as *no* runtime classes are allowed before this change. However, no user pods should be running mid-upgrade, and there is no plan to have system pods running in a runtime class.
```

Comment:
please walk through an actual scenario.  I *think* that if you require certain runtime classes for an SCC to be valid, when that list of required runtimeclasses disappears on a downgrade, the SCC will be left in place and suddenly grant access to a potentially increased list of powers.

For instance, if you had a kata SCC that allows untrusted users to run as root for kata containers, then downgraded, the existing SCC admission would allow untrusted users to run as root in all pods.

Reply:
Okay I will add that case. Better docs is my best idea to mitigate this case, though I'm open to others.